pkg/apis/batch/validation/validation.go

```diff
@@ -546,6 +546,19 @@ func ValidateJobTemplateSpec(spec *batch.JobTemplateSpec, fldPath *field.Path, o
 	return allErrs
 }
 
+func validateIndexedJobCompletionsMutatedWithParallelism(spec, oldSpec batch.JobSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if spec.CompletionMode != nil && *spec.CompletionMode != batch.IndexedCompletion {
+		return allErrs
+	}
```

Comment:
this makes `completions` completely mutable for non-indexed jobs (even able to be changed from a set value to nil). that doesn't match what is described in the KEP. I expected something more like this:

```go
// CompletionMode is defaulted, we can count on it being non-nil
if *spec.CompletionMode != batch.IndexedCompletion {
  // only indexed jobs are allowed to change completions
  return apivalidation.ValidateImmutableField(spec.Completions, oldSpec.Completions, fldPath.Child("completions"))
}

// Parallelism is defaulted, we can count on it being non-nil.
// Completions is not defaulted, so guard against nil (which is caught by validateJobSpec)
// A non-nil value for completions must match spec.Parallelism
if spec.Completions != nil && *oldSpec.Completions != *spec.Completions && *spec.Parallelism != *spec.Completions {
  allErrs = append(allErrs, field.Invalid(fldPath, spec.Completions, ...
}
```

Reply:
also add unit tests for:
* completions set → nil (disallow, don't panic)
* completions unchanged, parallelism lower to less than completions
* completions unchanged, parallelism raise to more than completions

Reply:
Ok I updated the validation logic here accordingly and added the additional unit tests described above. However, the logic became a bit more complicated than your example above because in order to get all the existing test cases to pass as well. Let me know if you have any thoughts.